### ISSUE TITLE
Support filtered retrieval from a topic by message key

### ISSUE THREAD

**Issue by jfallows** [reaktivity/nukleus-kafka.java#5]
Allows for more flexible mappings of logical message streams to physical Kafka topics, rather than forcing a one-to-one mapping.

**Comment by cmebarrow** [comment]
To support the last requirement (-1 to be a valid value for fetchKeyHash) we will need slightly different idl. One possibility would be:
```
struct KafkaBeginEx
{
    string16 topicName;
    varint64[] fetchOffsets;
    int32 fetchKeySize = -1;
    octets[fetchKeySize] fetchKey;
    int8 fetchKeyHashCount;
    int32[fetchKeyHashCount] fetchKeyHash = null;
}
```
It is an error to supply more than one fetchKeyHash.

**Comment by cmebarrow** [comment]
The offset we report for each filtered message should be the highest one we can report, not necessarily the offset of the message itself. For example if we fetch 10 messages, only the first one matches the logical topic, we should report the offset of the 10th when delivering the first message, in order to avoid unnecessary historical message retrieval if the client unsubscribed after receiving the first message then resubscribe later to resume consuming messages after the first message.

**Comment by jfallows** [comment]
Agreed, unless there is a way for us to send updated metadata describing the later fetch offset without implying an empty message.

**Comment by cmebarrow** [comment]
We should transmit this information as a data frame in the high-level Kafka stream even if it can't be exploited by the receiving protocol.

**Comment by jfallows** [comment]
Right, though probably using some new concept of `null` `DATA` frame, as opposed to empty `DATA` frame, which conveys transmission of an empty message that we don't want to imply here.

**Comment by cmebarrow** [comment]
When a fetchKey is specified (and optionally the fetchKeyHash) that implies messages will be fetched from only one partition. There must therefore be only one fetchOffset specified. Otherwise it is an error and the Begin frame will be rejected with a Reset.

**Comment by cmebarrow** [comment]
Test cases:

1. Client subscribes, kafka live fetch retrieves several messages, only the last one matches the fetch key and is delivered to the client. The offset reported on the DataEx is the high water mark of the record set
2. Client subscribes, kafka live fetch retrieves several messages, only the first matches and is delivered to the client. The offset reported on the DataEx is high watermark of the data set.
3 Client subscribes, kafka live fetch retrieves 3 messages, only the first and last are delivered to the client. The offset reported on the DataEx of the first message is the offset of the second (non-matching) message + 1,
